packages/gmail/src/Adaptor.js

```diff
@@ -0,0 +1,344 @@
+  });
+}
+
+async function getFileFromAttachment(attachment) {
+  const base64String = attachment?.data?.data;
+  if (!base64String) {
+    throw new Error('No data found in file.');
+  }
+
+  const fileContent = atob(base64String);
+
+  return isTesting ? fileContent.substring(0, 40) : fileContent;
+}
+
+async function getFileFromArchive(archive, filePattern) {
+  const base64String = archive?.data?.data;
+  if (!base64String) {
+    throw new Error('No data found in zip attachmentResponse.');
+  }
+
+  const compressedBuffer = Buffer.from(base64String, 'base64');
+  const directory = await unzipper.Open.buffer(compressedBuffer);
+
+  const file = directory?.files.find(f => filePattern.test(f.path));
+
+  if (!file) {
+    throw new Error('File not found in the archive.');
+  }
+
+  const fileBuffer = await file.buffer();
+  const fileString = fileBuffer.toString('base64');
+  const fileContent = atob(fileString);
+
+  return {
+    content: isTesting ? fileContent.substring(0, 40) : fileContent,
+    filename: file.path,
+  };
+}
+
+function getAttachmentInfo(messageResponse, regex) {
+  const parts = messageResponse?.data?.payload?.parts;
+  const part = parts?.find(p => regex.test(p.filename));
+
+  return part
+    ? { attachmentId: part.body.attachmentId, filename: part.filename }
+    : { attachmentId: null };
+}
+
+function getBodyFromMessage(fullMessage) {
+  const parts = fullMessage?.data?.payload?.parts;
+
+  const bodyPart = parts?.find(
+    part => part.mimeType === 'multipart/alternative'
+  );
+
+  const textBodyPart = bodyPart?.parts.find(
+    part => part.mimeType === 'text/plain'
+  );
+
+  const textBody = textBodyPart?.body?.data;
+  if (textBody) {
+    let body = Buffer.from(textBody, 'base64').toString('utf-8');
+    return isTesting ? body.substring(0, 40) : body;
+  }
+
+  return null;
+}
+
+/**
+ * Execute a sequence of oper.
+ * Wraps `language-common/execute`, and prepends initial state for http.
+ * @example
+ * execute(
+ *   create('foo'),
+ *   delete('bar')
+ * )(state)
+ * @private
+ * @param {Operations} operations - Operations to be performed.
+ * @returns {Operation}
+ */
+export function execute(...operations) {
+  const initialState = {
+    references: [],
+    data: null,
+  };
+
+  return state => {
+    // Note: we no longer need `steps` anymore since `commonExecute`
+    // takes each operation as an argument.
+    return commonExecute(
+      createConnection,
+      ...operations,
+      removeConnection
+    )({
+      ...initialState,
+      ...state,
+      configuration: normalizeOauthConfig(state.configuration),
+    });
+  };
+}
+
+function createConnection(state) {
+  const { access_token } = state.configuration;
+
+  const auth = new google.auth.OAuth2();
+  auth.credentials = { access_token };
+
+  gmail = google.gmail({ version: 'v1', auth });
+
+  return state;
+}
+
+function removeConnection(state) {
+  gmail = undefined;
+  return state;
+}
+
+function logError(error) {
```

Comment:
This function is currently unused - is that right? I would remove it before 1.0 if you don't need it